source/cpu_usage_sampler.c

```diff
@@ -0,0 +1,238 @@
+#    include <sys/sysinfo.h>
+#    include <sys/types.h>
+#endif
+
+/*********************************************************************************************************************
+ * Base implementation
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler_vtable {
+    int (*aws_get_cpu_sample_fn)(struct aws_cpu_sampler *sampler, double *output);
+    void (*aws_cpu_sampler_destroy)(struct aws_cpu_sampler *sampler);
+};
+
+struct aws_cpu_sampler {
+    const struct aws_cpu_sampler_vtable *vtable;
+    struct aws_allocator *allocator;
+    void *impl;
+};
+
+static void aws_cpu_sampler_destroy_default(struct aws_cpu_sampler *sampler) {
+    if (sampler == NULL) {
+        return;
+    }
+    aws_mem_release(sampler->allocator, sampler);
+}
+
+static struct aws_cpu_sampler_vtable aws_cpu_sampler_vtable_default = {
+    .aws_get_cpu_sample_fn = NULL,
+    .aws_cpu_sampler_destroy = aws_cpu_sampler_destroy_default,
+};
+
+/*********************************************************************************************************************
+ * Linux
+ ********************************************************************************************************************/
+
+#if defined(__linux__) || defined(__unix__)
+
+struct aws_cpu_sampler_linux {
+    struct aws_cpu_sampler base;
+
+    uint64_t cpu_last_total_user;
+    uint64_t cpu_last_total_user_low;
+    uint64_t cpu_last_total_system;
+    uint64_t cpu_last_total_idle;
+    // CPU usage is reported in deltas on Linux.
+    // For the first report we have to process and cache but not send.
+    bool cpu_only_gather;
+};
+
+static void s_get_cpu_usage_linux(
+    uint64_t *total_user,
+    uint64_t *total_user_low,
+    uint64_t *total_system,
+    uint64_t *total_idle) {
+
+    FILE *file;
+    int matched_results;
+    file = fopen("/proc/stat", "r");
+    matched_results = fscanf(
+        file,
+        "cpu %llu %llu %llu %llu",
+        (long long unsigned int *)total_user,
+        (long long unsigned int *)total_user_low,
+        (long long unsigned int *)total_system,
+        (long long unsigned int *)total_idle);
+    fclose(file);
+    if (matched_results == EOF || matched_results != 4) {
+        aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+}
+
+static int aws_get_cpu_sample_fn_linux(struct aws_cpu_sampler *sampler, double *output) {
+    struct aws_cpu_sampler_linux *sampler_linux = sampler->impl;
+
+    int return_result = AWS_OP_ERR;
+    uint64_t total_user, total_user_low, total_system, total_idle, total;
+    s_get_cpu_usage_linux(&total_user, &total_user_low, &total_system, &total_idle);
+    // total_combined needs to be double to allow for fractions
+    double percent, total_combined;
+
+    // If the value has overflown and wrapped around, swap the variables
+    uint64_t tmp;
+    if (total_user < sampler_linux->cpu_last_total_user) {
```

Comment:
let's make a helper compute-difference function that does this overflow check and adjust for arbitrary (start, end) uint64 pairs.

Also I don't think we want to swap.  The overflow case should be handled by:

```
if (overflow) {
    diff = (AWS_UINT64_MAX - start) + end;
} else {
    diff = end - start;
}
```